codalab/rest/interpret.py

```diff
@@ -183,6 +186,12 @@ def fetch_interpreted_worksheet(uuid):
     # Fetch items.
     worksheet_info['raw'] = get_worksheet_lines(worksheet_info)
 
+    # Replace searches with raw items.
+    # This needs to be done before get_worksheet_lines because this replaces
+    # user-written raw items.
+    worksheet_info['items'] = list(chain.from_iterable(
+        [expand_raw_item(raw_item) for raw_item in worksheet_info['items']]))
```

Comment:
This really worries me...`items` is now a list of a list of items?  It'd be nice if `items` really just corresponded to the raw items.  I thought the search directives would only influence the blocks.

Reply:
No, the chain.from_iterable flattens the list of list of items from the list comprehension into a list of items. Before I changed anything worksheet_info['items'] was populated with raw items, and then was replaced with the interpreted items. I can make it so that this is completely separated into worksheet_info['raw_items'], worksheet_info['replaced_raw_items'], and worksheet_info['interpreted_blocks'], if that makes more sense. 

I think this actually fits really nicely into how our parser/interpreter works currently, because a search is an alias for a list of bundles (or worksheets, in the case of wsearch), and this means that we don't have to include extra logic for every single block on how to handle searches.

Reply:
Well, one question is whether we want the conversion from items to blocks to be oblivious of search directives.  Certainly doing this replacement simplifies the blocks logic, but it does introduce 3 objects to reason about (raw items, replaced items, and blocks) rather than 2.  Also, it precludes us from later trying to lazily resolve the search directives during interpretation time rather than all at once.

Reply:
I don't think it gives us 3 objects to reason about, since replaced items are identical to raw items from the perspective of the raw -> block interpreter. 

Also, as we discussed before, search directives are important to the codalab architecture, but also used very sparsely, mostly in the homepage and to perform debugging. Therefore, I think it would be much more efficient to add logic for lazily resolving other directives, such as image and table, since that functionality would apply to all worksheets. Also, those gains would be realized for searches, since searches resolve into table blocks anyway.

Reply:
As long as only raw items and blocks are exposed to the frontend, then I'm happy.  If the replaced items can be encapsulated in the raw items => blocks conversion (even though we do break it up further, which is great)...

I don't understand your point about searches - can you please rephrase?

Are we committing to resolving all the search directives whenever we fetch a worksheet?  This seems to make certain things off limits - like being able to lazily fetch a very long list of bundles or worksheets dynamically...  Maybe it's fine to do that, but it seems worth thinking through the consequences.

Reply:
Yes, currently the frontend is completely unaware of the raw items. It renders purely based on the blocks. 

Sorry for being unclear; I think what I was trying to say is that the plan is already to lazily resolve the contents of display directives that require more than just the bundle metadata, such as images, contents, and genpaths in tables. These all require disk reads, and are what contribute most to the slow load times of large and complex worksheets. For example, if a table of 10 items wants the first line of the file results.txt, that will hit the file system 10 separate times. I was under the impression that this refactor is laying the grounds for lazily loading these disk-intensive tasks. 

Meanwhile, search is definitely not a common tool for most worksheets, it is mostly used as a diagnostic tool for us to build dashboards such as the current homepage. Additionally, it will gain the same performance gains that I described above. 

We are committing to resolving all of the search directives whenever we fetch a worksheet, but that was already being done previously, so it wouldn't make anything slower than it already is now. I'm having a hard time thinking of a use case for lazily fetching a very long (>1000) list of bundles dynamically.